### ISSUE TITLE
How long to train this model

### ISSUE THREAD

**Issue by kikirizki** [nii-yamagishilab/project-NN-Pytorch-scripts#11]
Thank you for your amazing work. Btw how long it take to train this model

**Comment by TonyWangX** [comment]
Are you asking about the absolute time per epoch (e.g, 1 hour per epoch)? 

It depends on the dataset, GPU card, and so on.

For the pre-trained NSF models, there is a training log in each __pre-trained folder, which was produced when I trained the model on our server.

For example, https://github.com/nii-yamagishilab/project-NN-Pytorch-scripts/blob/master/project/01-nsf/cyc-noise-nsf-4/__pre_trained/log_train

Given the training set with 198395040 waveform samples (i.e, 3.44 hours in duration for 16kHz wavefom) and a development set with 13548720 waveform samples (i.e., 0.23 hours), the model needs ~2600 s for one epoch on the server with ... (Sorry I didn't write the GPU card into the log, but I remembered that it was Nvidia V100 32GB). 

So, less than one hour per epoch on this ~4 hours dataset. The default training will take 50 epochs, thus the total training time will be around 36 hours.

**Comment by TonyWangX** [comment]
The format of log file is quite old.  
If you train the model using the current scripts, the log will includes the GPU card name.

**Comment by TonyWangX** [comment]
**One point to increase the speed**

The script uses cudnn_deterministic = True and cudnn_benckmark = False by default in order to get reproducible results (i.e., same input, same network weights, same output https://pytorch.org/docs/stable/notes/randomness.html), 
https://github.com/nii-yamagishilab/project-NN-Pytorch-scripts/blob/379ef72fcee3b9ea3e1ba28268479a56cbf63c3a/core_scripts/startup_config.py#L43
However, these two options greatly affect the training speed.


**To increase the speed**
Set cudnn_deterministic = False and cudnn_benckmark = True in startup_config.py


In addition to the way above, I also prepared command arguments to toggle the two options. However, this needs a slight change of the main.py:
1. change this function to nii_startup.set_random_seed(args.seed, args)
https://github.com/nii-yamagishilab/project-NN-Pytorch-scripts/blob/379ef72fcee3b9ea3e1ba28268479a56cbf63c3a/project/01-nsf/cyc-noise-nsf-4/main.py#L46
2. when running the script, use options $: python main.py --cudnn-deterministic-toggle --cudnn-benchmark-toggle

**Comment by TonyWangX** [comment]
Another way to optimize the speed needs more work.

For example, Pytorch conv1d API assumes tensor in (batch, channel, length) format. 
But I am more comfortable with (batch, length, dim) format and added "permute" around conv1d and other APIs.
https://github.com/nii-yamagishilab/project-NN-Pytorch-scripts/blob/379ef72fcee3b9ea3e1ba28268479a56cbf63c3a/sandbox/block_nn.py#L94

This is convenient for me when I re-use the modules. But these additional operations increase the overhead.
I didn't do (and I am not good at) code optimization for speed.

If you want to optimize the speed, I think this is the place to hack

**Comment by kikirizki** [comment]
Wow thank you so much for your responses that cleared up my confusion
